### ISSUE TITLE
Grafana 9.1.4 NoData reduce problem

### ISSUE THREAD

**Issue by matrog** [grafana/grafana#55085]
After upgrading from grafana 8.5.6 to 9.1.4 i'm getting the following error message if the query returns "NoData", which is a totally expected situation in case we do not receive any data from one of our probes.

<img width="459" alt="3" src="https://user-images.githubusercontent.com/103855/189898999-e01616c7-ec9e-4752-bcfe-f66827ecfbbb.png">

This alert used to work fine in v8.5.6 producing a NoData Alert.

Here's the alert config

<img width="1511" alt="1" src="https://user-images.githubusercontent.com/103855/189899111-136e207d-c5b4-4014-84f1-e75ddc97b25b.png">

then a reduce and a math expression

<img width="1446" alt="2" src="https://user-images.githubusercontent.com/103855/189899158-bc64c651-00fa-4f05-b7b3-18071f94c316.png">

Log Snippet

```
logger=ngalert uid=CwG7uaM4k org=1 version=1 attempt=0 now=2022-09-13T12:22:00Z t=2022-09-13T12:22:06.93425422Z level=error msg="failed
 to evaluate alert rule" results="[{Instance: State:Error Error:failed to execute conditions: can only reduce type series, got type noD
ata EvaluatedAt:2022-09-13 12:22:00 +0000 UTC EvaluationDuration:6.934249393s EvaluationString: Values:map[]}]" duration=25.161092ms
logger=ngalert t=2022-09-13T12:22:06.934335965Z level=debug msg="state manager processing evaluation results" uid=CwG7uaM4k resultCount
=1
```
The expected result is NoData state if the query returns NoData

**Environment**:
- Grafana version: 9.1.4
- Data source type & version: Flux (with Influxdb 2.1.0)
- OS Grafana is installed on: RedHat 8.6
- User OS & Browser: Chrome / Firefox on macOS
- Grafana plugins: N/A
- Others: N/A

**Comment by majkinetor** [comment]
I have the same problem after upgrade.

Even if I set:

![image](https://user-images.githubusercontent.com/85767/190422243-4289e3ec-ae10-4150-ba11-033e930fe987.png)


I get OK annotations, so it transitions from OK to OK which is certainly a bug.

I will have to disable all alerts until this is resolved since OK/NoData/Errrors spam like there is no tomorrow so meaningfull alerts are lost in the noise.

**Comment by jl-678** [comment]
I see this same problem too and [posted about it including a screenshot ](https://community.grafana.com/t/getting-rid-of-nodata-markers/72418)in the Grafana Forum.

**Comment by grobinson-grafana** [comment]
Hi! 👋

This is an issue with how Grafana reduces data frames from InfluxDB. With InfluxDB, no data returns a `mathexp.NoData` data frame, where as with Prometheus no data returns no data frames. Grafana thinks there is data and attempts to reduce the `mathexpr.NoData` data frame, but `mathexpr.NoData` cannot be reduced and so Grafana returns the error `can only reduce type series, got type noData`.

**Comment by jl-678** [comment]
Thank you for the explanation.  Am I correct in interpreting your answer to say that this is indeed a bug in how Grafana treats no data from Influx?

**Comment by kylebrandt** [comment]
I think we need to go the other way here and keep the single frame -- targeting that by v10 (data plane contract -- https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/contract_docs/contract.md#no-data-and-empty).

The primary reason for this is Frame Metadata, in particular we need it for information returned by the backend about queries that were made:
 - ExecutedQueryString (post "raw/text query" macro expansion in various datasources)
 - More future - the type identifier, so we in some cases we may know that 

(Then reduce should handle NoData inputs that have a frame attached instead of erroring on them). Been a bit since I thought about that so not sure of exact behaviors, But NoData into a reduce probably also returns NoData (and then math probably needs to handle it to).

**Comment by grobinson-grafana** [comment]
Hi! I have opened a pull request to fix this https://github.com/grafana/grafana/pull/55347. It should be available in Grafana 9.1.6.

**Comment by grobinson-grafana** [comment]
Hi! I have opened a pull request to fix this https://github.com/grafana/grafana/pull/55347. It should be available in Grafana 9.1.6.

**Comment by grobinson-grafana** [comment]
Hi! I don't think https://github.com/grafana/grafana/pull/55347 will be merged in time for 9.1.6, so it will instead be available in Grafana 9.1.7.

**Comment by vinivosh** [comment]
I don't understand how #55347 fixed this, actually. I'm on v9.2.0-beta.1 right now and the `can only reduce type series, got type noData` error is still spat out when an InfluxDB query of mine returns nothing at all. This happens both in a "Reduce" condition or in the "Classic Condition".

Grafana then just sends a DatasourceError notification, instead of dealing with a NoData situation correctly. This didn't happen in Grafana 7.

Specially annoying when you have other queries in the same alert. They won't be run even if they'd return something, just because one other query returned nothing.

Am I missing something here?

**Comment by matrog** [comment]
i just installed 9.1.7 and I can confirm the issue is solved, now the NoData is handled in proper way

**Comment by grobinson-grafana** [comment]
Hi @vinivosh! :wave: Would it be possible to see the SQL query and the database schema of the table you are querying in your alert rule?

**Comment by vinivosh** [comment]
> 

I'm actually using Influx' own Flux query language to query data from InfluxDB v2.3.0.

The query in question:
```
from(bucketID: v.defaultBucket)
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["device_id"] == "BBHZJAP")
  |> filter(fn: (r) => r["channel"] == "0")
  |> filter(fn: (r) => r["_measurement"] == "temperature")
  |> aggregateWindow(every: v.windowPeriod, fn: last)
  |> map(fn: (r) => ({ _value:r._value, _time:r._time, _field:"Temperature Street X - Device Y" }))
```

It currently returns nothing, because the hardware isn't set up properly at the moment. I still wanted other queries in the alert to be ran, though, as they do return data. Or at least for Grafana to deal with this scenario normally, by sending a NoData alert.

The conditions on the alert:
<img width="802" alt="image" src="https://user-images.githubusercontent.com/36175157/194964001-5d6c7e0d-f728-4da3-a255-7a623448fbf9.png">

I also tried using Math or Reduce conditions, but the same DatasourceError happens.

**Comment by grobinson-grafana** [comment]
Hi @vinivosh! I'm not sure I can reproduce the issue. I copied the FluxQL and created an alert with Reduce, and I get a `DatasourceNoData` alert. I can reproduce it with Classic Condition but that's because we haven't applied the same fix to classic conditions (should have a fix up this week).

**Comment by matrog** [comment]
I did some tests and i'm getting same error as @vinivosh , I was not able to experience that before because I don't use classic condition but always the reduce.

<img width="1696" alt="Screenshot 2022-10-11 at 14 50 22" src="https://user-images.githubusercontent.com/103855/195095476-73474838-de11-4645-a625-054d6d87ec0c.png">

with the same query i'm stil getting the error

**Comment by grobinson-grafana** [comment]
Hi @matrog! In the screenshot you are using a classic condition which hasn't been fixed. It's just Reduce expressions that have been fixed, but @vinivosh mentioned he is still getting the error in Reduce expressions.

**Comment by matrog** [comment]
@grobinson-grafana is there an issue to track the status of the classic condition + NoData?

**Comment by grobinson-grafana** [comment]
Hi @matrog! I've just merged a fix for NoData in Classic Conditions which should be available in both 9.1.9 and 9.2.1.

**Comment by matrog** [comment]
Super! thanks
